File: s6-optional-challenges/frequency_counter/same_frequency/solution.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	fmt.Println(sameFrequency(182, 281))         // true
	fmt.Println(sameFrequency(34, 14))           // false
	fmt.Println(sameFrequency(3589578, 5879385)) // true
	fmt.Println(sameFrequency(22, 222))          // false
}

func sameFrequency(n1, n2 int) bool {
	s1 := strconv.Itoa(n1)
	s2 := strconv.Itoa(n2)

	if len(s1) != len(s2) {
		return false
	}

	lookup := make(map[rune]int)
	for _, l := range s1 {
		count := lookup[l]
		lookup[l] = count + 1
	}

	for _, l := range s2 {
		if val, _ := lookup[l]; val == 0 {
			return false
		}
		lookup[l]--
	}
	return true
}

File: s6-optional-challenges/frequency_counter/same_frequency/solution.py
def sameFrequency(n1, n2):
    s1 = str(n1)
    s2 = str(n2)

    if len(s1) != len(s2):
        return False

    lookup = {}
    for l in s1:
        if l in lookup:
            lookup[l] += 1
        else:
            lookup[l] = 1

    for l in s2:
        if l not in lookup or lookup[l] == 0:
            return False
        lookup[l] -= 1

    return True


print(sameFrequency(182, 281))  # true
print(sameFrequency(34, 14))  # false
print(sameFrequency(3589578, 5879385))  # true
print(sameFrequency(22, 222))  # false

